presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java

```diff
@@ -741,15 +758,103 @@ public ParquetStructConverter(Type prestoType, String columnName, GroupType entr
                     prestoTypeParameters.size(),
                     fieldTypes.size());
 
-            this.rowType = prestoType;
-            this.fieldIndex = fieldIndex;
+            ImmutableList.Builder<BlockFieldConverter> converters = ImmutableList.builder();
+            for (int i = 0, n = prestoTypeParameters.size(); i < n; i++) {
+                parquet.schema.Type fieldType = fieldTypes.get(i);
+                converters.add(createConverter(prestoTypeParameters.get(i), columnName + "." + fieldType.getName(), fieldType, i, false));
+            }
+            return converters.build();
+        }
+
+        private List<BlockFieldConverter> createConvertersByName(String columnName, List<RowField> prestoTypeFields, List<parquet.schema.Type> fieldTypes)
+        {
+            int fieldCount = fieldTypes.size();
+            int prestoFieldCount = prestoTypeFields.size();
+            if (fieldCount != prestoFieldCount) {
+                log.info("Schema mismatch, metastore schema for row column %s has %s fields but parquet schema has %s fields",
+                        columnName,
+                        prestoFieldCount,
+                        fieldCount);
+            }
 
-            ImmutableList.Builder<BlockConverter> converters = ImmutableList.builder();
-            for (int i = 0; i < prestoTypeParameters.size(); i++) {
+            Map<String, ObjectPair<Integer, Type>> fieldTypeMap = new HashMap<String, ObjectPair<Integer, Type>>(prestoFieldCount);
+            for (int i = 0, n = prestoTypeFields.size(); i < n; i++) {
+                RowField field = prestoTypeFields.get(i);
+                fieldTypeMap.put(field.getName().get().toLowerCase(), ObjectPair.create(Integer.valueOf(i), field.getType()));
+            }
+
+            boolean outOfOrder = false;
```

Comment:
let's add some comments here
when reading parquet files, we need to pass converters and parquet types in parquet schema fields order
after getting the parquet values, we need to build presto blocks in metastore schema fields order